### PR TITLE
[BEAM-11116] Make Java SDK container pull licenses without gradle subprocess

### PR DESCRIPTION
**Pull request by kennknowles**
Before:

 - `:sdks:java:container:pullLicenses` runs a shell script that invokes Gradle at the top level as a subprocess
 - no artifacts are shared
 - always has `--rerun-tasks` so it also reruns everything
 - the two gradle processes overwrite and clobber/corrupt the same artifacts sometimes
 - seems to do a full java compile of some modules as part of the subprocess (sometimes?) which takes 7.5 minutes on my laptop
 - finds 80 license files and 22 notice files

After:

 - artifacts are shared
 - intermediate outputs are cached
 - no more clobbering
 - from a clean state takes 1 minute (most of it gradle configuration time) on my laptop
 - finds 182 license files and 41 notice files, and `diff` shows it is a strict superset (hence safe)
    - some of these appear to be duplicate pulls, but not all
    - before and after, this is entirely driven by what deps the `generateLicenseReport` task thinks matter
    - before it was counting on a global run to cause projects to get configured; after, it is explicitly configuring the project we need. I think afterwards it is more accurate and deterministic, and ready to be refined.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kennknowles** [comment]
Pulling licenses fails with this change. I haven't figured out what made the difference. But it is actually failing silently on master. The license that cannot be retrieved is being silently omitted on master.

**Comment by kennknowles** [comment]
In fact, attempting an apples-to-apples comparison (had to `gunzip` some things and choose directories) we see via `diff -qur build-orig/target/third_party_licenses/ build-new/java_third_party_licenses` there are serious problems of many licenses not getting pulled on master:

```
Files build-orig/target/third_party_licenses/api-common-1.8.1.jar/LICENSE and build-new/java_third_party_licenses/api-common-1.8.1.jar/LICENSE differ
Files build-orig/target/third_party_licenses/args4j-2.33.jar/LICENSE and build-new/java_third_party_licenses/args4j-2.33.jar/LICENSE differ
Only in build-new/java_third_party_licenses/avro-1.8.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/avro-1.8.2.jar: META-INF
Only in build-new/java_third_party_licenses/bcel-6.5.0.jar: LICENSE
Only in build-orig/target/third_party_licenses/bcel-6.5.0.jar: META-INF
Only in build-new/java_third_party_licenses/beam-vendor-bytebuddy-1_10_8-0.1.jar: LICENSE
Only in build-orig/target/third_party_licenses/beam-vendor-bytebuddy-1_10_8-0.1.jar: META-INF
Only in build-new/java_third_party_licenses/beam-vendor-grpc-1_26_0-0.3.jar: LICENSE
Only in build-orig/target/third_party_licenses/beam-vendor-grpc-1_26_0-0.3.jar: LICENSE.txt
Only in build-orig/target/third_party_licenses/beam-vendor-grpc-1_26_0-0.3.jar: META-INF
Only in build-orig/target/third_party_licenses/beam-vendor-grpc-1_26_0-0.3.jar: certs
Only in build-new/java_third_party_licenses: beam_java_dependency_list.csv
Only in build-orig/target/third_party_licenses/: beam_java_dependency_list.csv.gz
Only in build-new/java_third_party_licenses/checker-qual-3.7.0.jar: LICENSE
Only in build-orig/target/third_party_licenses/checker-qual-3.7.0.jar: META-INF
Only in build-new/java_third_party_licenses/commons-beanutils-1.9.3.jar: LICENSE
Only in build-orig/target/third_party_licenses/commons-beanutils-1.9.3.jar: META-INF
Only in build-new/java_third_party_licenses/commons-codec-1.14.jar: LICENSE
Only in build-orig/target/third_party_licenses/commons-codec-1.14.jar: META-INF
Only in build-new/java_third_party_licenses/commons-collections-3.2.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/commons-collections-3.2.2.jar: META-INF
Only in build-new/java_third_party_licenses/commons-compress-1.20.jar: LICENSE
Only in build-orig/target/third_party_licenses/commons-compress-1.20.jar: META-INF
Only in build-new/java_third_party_licenses/commons-lang3-3.9.jar: LICENSE
Only in build-orig/target/third_party_licenses/commons-lang3-3.9.jar: META-INF
Only in build-new/java_third_party_licenses/commons-logging-1.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/commons-logging-1.2.jar: META-INF
Only in build-new/java_third_party_licenses/commons-math3-3.6.1.jar: LICENSE
Only in build-orig/target/third_party_licenses/commons-math3-3.6.1.jar: META-INF
Only in build-new/java_third_party_licenses/commons-text-1.8.jar: LICENSE
Only in build-orig/target/third_party_licenses/commons-text-1.8.jar: META-INF
Files build-orig/target/third_party_licenses/dom4j-2.1.3.jar/LICENSE and build-new/java_third_party_licenses/dom4j-2.1.3.jar/LICENSE differ
Files build-orig/target/third_party_licenses/gax-1.57.1.jar/LICENSE and build-new/java_third_party_licenses/gax-1.57.1.jar/LICENSE differ
Files build-orig/target/third_party_licenses/gax-httpjson-0.71.0.jar/LICENSE and build-new/java_third_party_licenses/gax-httpjson-0.71.0.jar/LICENSE differ
Only in build-new/java_third_party_licenses/hamcrest-core-2.1.jar: LICENSE
Only in build-orig/target/third_party_licenses/hamcrest-core-2.1.jar: README.txt
Only in build-new/java_third_party_licenses/hamcrest-library-2.1.jar: LICENSE
Only in build-orig/target/third_party_licenses/hamcrest-library-2.1.jar: README.txt
Only in build-new/java_third_party_licenses/httpclient-4.5.10.jar: LICENSE
Only in build-orig/target/third_party_licenses/httpclient-4.5.10.jar: META-INF
Only in build-new/java_third_party_licenses/httpcore-4.4.12.jar: LICENSE
Only in build-orig/target/third_party_licenses/httpcore-4.4.12.jar: META-INF
Files build-orig/target/third_party_licenses/icu4j-63.1.jar/LICENSE and build-new/java_third_party_licenses/icu4j-63.1.jar/LICENSE differ
Only in build-new/java_third_party_licenses/jackson-annotations-2.10.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/jackson-annotations-2.10.2.jar: META-INF
Only in build-new/java_third_party_licenses/jackson-core-2.10.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/jackson-core-2.10.2.jar: META-INF
Only in build-new/java_third_party_licenses/jackson-core-asl-1.9.13.jar: LICENSE
Only in build-orig/target/third_party_licenses/jackson-core-asl-1.9.13.jar: META-INF
Only in build-new/java_third_party_licenses/jackson-databind-2.10.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/jackson-databind-2.10.2.jar: META-INF
Only in build-new/java_third_party_licenses/jackson-dataformat-yaml-2.10.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/jackson-dataformat-yaml-2.10.2.jar: META-INF
Only in build-new/java_third_party_licenses/jackson-mapper-asl-1.9.13.jar: LICENSE
Only in build-orig/target/third_party_licenses/jackson-mapper-asl-1.9.13.jar: META-INF
Only in build-new/java_third_party_licenses/javaruntype-1.3.jar: LICENSE
Only in build-orig/target/third_party_licenses/javaruntype-1.3.jar: META-INF
Only in build-new/java_third_party_licenses/javax.annotation-api-1.3.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/javax.annotation-api-1.3.2.jar: META-INF
Only in build-orig/target/third_party_licenses/jaxen-1.1.6.jar: META-INF
Only in build-new/java_third_party_licenses/joda-time-2.10.5.jar: LICENSE
Only in build-orig/target/third_party_licenses/joda-time-2.10.5.jar: META-INF
Files build-orig/target/third_party_licenses/kafka-clients-1.0.0.jar/LICENSE and build-new/java_third_party_licenses/kafka-clients-1.0.0.jar/LICENSE differ
Only in build-orig/target/third_party_licenses/kafka-clients-1.0.0.jar: NOTICE
Files build-orig/target/third_party_licenses/kafka-clients-2.1.0.jar/LICENSE and build-new/java_third_party_licenses/kafka-clients-2.1.0.jar/LICENSE differ
Only in build-orig/target/third_party_licenses/kafka-clients-2.1.0.jar: NOTICE
Only in build-new/java_third_party_licenses/log4j-api-2.6.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/log4j-api-2.6.2.jar: META-INF
Files build-orig/target/third_party_licenses/mockito-core-3.0.0.jar/LICENSE and build-new/java_third_party_licenses/mockito-core-3.0.0.jar/LICENSE differ
Only in build-new/java_third_party_licenses/objenesis-3.0.1.jar: LICENSE
Only in build-orig/target/third_party_licenses/objenesis-3.0.1.jar: META-INF
Only in build-new/java_third_party_licenses/postgresql-42.2.16.jar: LICENSE
Only in build-orig/target/third_party_licenses/postgresql-42.2.16.jar: META-INF
Only in build-new/java_third_party_licenses/powermock-api-mockito2-2.0.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/powermock-api-mockito2-2.0.2.jar: META-INF
Only in build-new/java_third_party_licenses/powermock-api-support-2.0.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/powermock-api-support-2.0.2.jar: META-INF
Only in build-new/java_third_party_licenses/powermock-core-2.0.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/powermock-core-2.0.2.jar: META-INF
Only in build-new/java_third_party_licenses/powermock-module-junit4-2.0.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/powermock-module-junit4-2.0.2.jar: META-INF
Only in build-new/java_third_party_licenses/powermock-module-junit4-common-2.0.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/powermock-module-junit4-common-2.0.2.jar: META-INF
Only in build-new/java_third_party_licenses/powermock-reflect-2.0.2.jar: LICENSE
Only in build-orig/target/third_party_licenses/powermock-reflect-2.0.2.jar: META-INF
Only in build-new/java_third_party_licenses/snappy-java-1.1.4.jar: LICENSE
Only in build-orig/target/third_party_licenses/snappy-java-1.1.4.jar: org
Only in build-new/java_third_party_licenses/spotbugs-4.0.6.jar: LICENSE
Only in build-orig/target/third_party_licenses/spotbugs-4.0.6.jar: edu
Only in build-new/java_third_party_licenses/spring-core-4.3.18.RELEASE.jar: LICENSE
Only in build-orig/target/third_party_licenses/spring-core-4.3.18.RELEASE.jar: META-INF
Only in build-new/java_third_party_licenses/spring-expression-4.3.18.RELEASE.jar: LICENSE
Only in build-orig/target/third_party_licenses/spring-expression-4.3.18.RELEASE.jar: META-INF
Only in build-new/java_third_party_licenses/threetenbp-1.4.4.jar: LICENSE
Only in build-orig/target/third_party_licenses/threetenbp-1.4.4.jar: META-INF
```

**Comment by kennknowles** [comment]
To summarize the licenses that are pulled _only_ after this change: `diff -qur build-orig/target/third_party_licenses/ build-new/java_third_party_licenses | grep 'Only in build-new' | grep -i LICENSE`

```
Only in build-new/java_third_party_licenses/avro-1.8.2.jar: LICENSE
Only in build-new/java_third_party_licenses/bcel-6.5.0.jar: LICENSE
Only in build-new/java_third_party_licenses/beam-vendor-bytebuddy-1_10_8-0.1.jar: LICENSE
Only in build-new/java_third_party_licenses/beam-vendor-grpc-1_26_0-0.3.jar: LICENSE
Only in build-new/java_third_party_licenses: beam_java_dependency_list.csv
Only in build-new/java_third_party_licenses/checker-qual-3.7.0.jar: LICENSE
Only in build-new/java_third_party_licenses/commons-beanutils-1.9.3.jar: LICENSE
Only in build-new/java_third_party_licenses/commons-codec-1.14.jar: LICENSE
Only in build-new/java_third_party_licenses/commons-collections-3.2.2.jar: LICENSE
Only in build-new/java_third_party_licenses/commons-compress-1.20.jar: LICENSE
Only in build-new/java_third_party_licenses/commons-lang3-3.9.jar: LICENSE
Only in build-new/java_third_party_licenses/commons-logging-1.2.jar: LICENSE
Only in build-new/java_third_party_licenses/commons-math3-3.6.1.jar: LICENSE
Only in build-new/java_third_party_licenses/commons-text-1.8.jar: LICENSE
Only in build-new/java_third_party_licenses/hamcrest-core-2.1.jar: LICENSE
Only in build-new/java_third_party_licenses/hamcrest-library-2.1.jar: LICENSE
Only in build-new/java_third_party_licenses/httpclient-4.5.10.jar: LICENSE
Only in build-new/java_third_party_licenses/httpcore-4.4.12.jar: LICENSE
Only in build-new/java_third_party_licenses/jackson-annotations-2.10.2.jar: LICENSE
Only in build-new/java_third_party_licenses/jackson-core-2.10.2.jar: LICENSE
Only in build-new/java_third_party_licenses/jackson-core-asl-1.9.13.jar: LICENSE
Only in build-new/java_third_party_licenses/jackson-databind-2.10.2.jar: LICENSE
Only in build-new/java_third_party_licenses/jackson-dataformat-yaml-2.10.2.jar: LICENSE
Only in build-new/java_third_party_licenses/jackson-mapper-asl-1.9.13.jar: LICENSE
Only in build-new/java_third_party_licenses/javaruntype-1.3.jar: LICENSE
Only in build-new/java_third_party_licenses/javax.annotation-api-1.3.2.jar: LICENSE
Only in build-new/java_third_party_licenses/joda-time-2.10.5.jar: LICENSE
Only in build-new/java_third_party_licenses/log4j-api-2.6.2.jar: LICENSE
Only in build-new/java_third_party_licenses/objenesis-3.0.1.jar: LICENSE
Only in build-new/java_third_party_licenses/postgresql-42.2.16.jar: LICENSE
Only in build-new/java_third_party_licenses/powermock-api-mockito2-2.0.2.jar: LICENSE
Only in build-new/java_third_party_licenses/powermock-api-support-2.0.2.jar: LICENSE
Only in build-new/java_third_party_licenses/powermock-core-2.0.2.jar: LICENSE
Only in build-new/java_third_party_licenses/powermock-module-junit4-2.0.2.jar: LICENSE
Only in build-new/java_third_party_licenses/powermock-module-junit4-common-2.0.2.jar: LICENSE
Only in build-new/java_third_party_licenses/powermock-reflect-2.0.2.jar: LICENSE
Only in build-new/java_third_party_licenses/snappy-java-1.1.4.jar: LICENSE
Only in build-new/java_third_party_licenses/spotbugs-4.0.6.jar: LICENSE
Only in build-new/java_third_party_licenses/spring-core-4.3.18.RELEASE.jar: LICENSE
Only in build-new/java_third_party_licenses/spring-expression-4.3.18.RELEASE.jar: LICENSE
Only in build-new/java_third_party_licenses/threetenbp-1.4.4.jar: LICENSE
```

File: sdks/java/container/build.gradle [hunk 1]
```diff
@@ -24,6 +24,9 @@ plugins {
 applyGoNature()
 applyDockerNature()
 
+evaluationDependsOn(":sdks:java:harness")
```
Comment (kennknowles) [inline sdks/java/container/build.gradle:27]: I believe these lines are the key to not needing the global run and not needing `--rerun-tasks`

**Review by Hannah-Jiang — APPROVED** [review]

**Comment by Hannah-Jiang** [comment]
Thanks for improving it.

**Pull request merged**